### PR TITLE
Fix empty tuple handling in a few formats

### LINKED ISSUE #71583: Can't read empty tuple from Arrow file
**Describe the bug**
A logical error happens on the client if an empty tuple is imported from an Arrow file.

**How to reproduce**
Run:

```sql
CREATE TABLE t0 (c0 Tuple()) ENGINE = Memory();
INSERT INTO TABLE t0 (c0) VALUES (());
INSERT INTO TABLE FUNCTION file('<path to file>', 'Arrow', 'c0 Tuple()') SELECT c0 FROM t0;
INSERT INTO TABLE t0 (c0) FROM INFILE '<path to file>' FORMAT Arrow;
/*
Logical error: 'This function cannot be used to construct empty tuple. It is a bug'.
*/
```
Stack trace:
```
Logical error: 'This function cannot be used to construct empty tuple. It is a bug'.
Stack trace (when copying this message, always include the lines below):

0. std::exception::capture() @ 0x0000000019428f0b
1. contrib/llvm-project/libcxx/include/exception:114: Poco::Exception::Exception(String const&, int) @ 0x0000000038ca9dc5
2. src/Common/Exception.cpp:109: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x00000000255838ab
3. DB::Exception::Exception(PreformattedMessage&&, int) @ 0x00000000192a7296
4. DB::Exception::Exception<>(int, FormatStringHelperImpl<>) @ 0x00000000192c0cef
5. src/Columns/ColumnTuple.cpp:69: DB::ColumnTuple::create(std::vector<COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::allocator<COW<DB::IColumn>::immutable_ptr<DB::IColumn>>> const&) @ 0x000000003386152b
6. src/Columns/ColumnTuple.h:43: DB::readNonNullableColumnFromArrowColumn(std::shared_ptr<arrow::ChunkedArray> const&, String, std::unordered_map<String, DB::ArrowColumnToCHColumn::DictionaryInfo, std::hash<String>, std::equal_to<String>, std::allocator<std::pair<String const, DB::ArrowColumnToCHColumn::DictionaryInfo>>>, std::shared_ptr<DB::IDataType const>, bool, DB::ReadColumnFromArrowColumnSettings const&) @ 0x00000000353fb607
7. src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp:1142: DB::readColumnFromArrowColumn(std::shared_ptr<arrow::ChunkedArray> const&, String, std::unordered_map<String, DB::ArrowColumnToCHColumn::DictionaryInfo, std::hash<String>, std::equal_to<String>, std::allocator<std::pair<String const, DB::ArrowColumnToCHColumn::DictionaryInfo>>>, std::shared_ptr<DB::IDataType const>, bool, bool, DB::ReadColumnFromArrowColumnSettings const&) @ 0x00000000353eabbb
8. src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp:1344: DB::ArrowColumnToCHColumn::arrowColumnsToCHChunk(std::unordered_map<String, DB::ArrowColumnToCHColumn::ArrowColumn, std::hash<String>, std::equal_to<String>, std::allocator<std::pair<String const, DB::ArrowColumnToCHColumn::ArrowColumn>>> const&, unsigned long, DB::BlockMissingValues*) @ 0x00000000353ec703
9. src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp:1250: DB::ArrowColumnToCHColumn::arrowTableToCHChunk(std::shared_ptr<arrow::Table> const&, unsigned long, DB::BlockMissingValues*) @ 0x00000000353ebcb3
10. src/Processors/Formats/Impl/ArrowBlockInputFormat.cpp:92: DB::ArrowBlockInputFormat::read() @ 0x00000000353cd952
11. src/Processors/Formats/IInputFormat.cpp:19: DB::IInputFormat::generate() @ 0x00000000353b4f75
12. src/Processors/ISource.cpp:139: DB::ISource::tryGenerate() @ 0x0000000035374694
13. src/Processors/ISource.cpp:108: DB::ISource::work() @ 0x0000000035373cba
14. src/Processors/Executors/ExecutionThreadContext.cpp:47: DB::ExecutionThreadContext::executeTask() @ 0x00000000353a3e20
15. src/Processors/Executors/PipelineExecutor.cpp:289: DB::PipelineExecutor::executeStepImpl(unsigned long, std::atomic<bool>*) @ 0x000000003538eac3
16. src/Processors/Executors/PipelineExecutor.cpp:163: DB::PipelineExecutor::executeStep(std::atomic<bool>*) @ 0x000000003538e0a3
17. src/Processors/Executors/PullingPipelineExecutor.cpp:54: DB::PullingPipelineExecutor::pull(DB::Chunk&) @ 0x00000000353ab694
18. src/Storages/StorageFile.cpp:1474: DB::StorageFileSource::generate() @ 0x0000000033e7b59f
19. src/Processors/ISource.cpp:139: DB::ISource::tryGenerate() @ 0x0000000035374694
20. src/Processors/ISource.cpp:108: DB::ISource::work() @ 0x0000000035373cba
21. src/Processors/Executors/ExecutionThreadContext.cpp:47: DB::ExecutionThreadContext::executeTask() @ 0x00000000353a3e20
22. src/Processors/Executors/PipelineExecutor.cpp:289: DB::PipelineExecutor::executeStepImpl(unsigned long, std::atomic<bool>*) @ 0x000000003538eac3
23. src/Processors/Executors/PipelineExecutor.cpp:255: DB::PipelineExecutor::executeImpl(unsigned long, bool) @ 0x000000003538db36
24. src/Processors/Executors/PipelineExecutor.cpp:126: DB::PipelineExecutor::execute(unsigned long, bool) @ 0x000000003538d74a
25. src/Processors/Executors/PullingAsyncPipelineExecutor.cpp:83: void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<true, true>::ThreadFromGlobalPoolImpl<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&&)::'lambda'(), void ()>>(std::__function::__policy_storage const*) @ 0x00000000353a926c
26. contrib/llvm-project/libcxx/include/__functional/function.h:848: ? @ 0x00000000256ba7cf
27. contrib/llvm-project/libcxx/include/__functional/invoke.h:359: ? @ 0x00000000256c19ca
28. ? @ 0x0000717c200a1e2e
29. ? @ 0x0000717c20133a4c
```

### PR DESCRIPTION
**Pull request by al13n321**
### Changelog category (leave one):
- Bug Fix

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix handling of empty tuples in some input and output formats (e.g. Parquet, Arrow).

Closes https://github.com/ClickHouse/ClickHouse/issues/71583

When writing or reading empty tuples, a few formats were returning LOGICAL_ERROR, a couple formats were crashing. Here are the trivial fixes.

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit 0b3f66bdd7b0c2090626a15566050f9f6d8351f2 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/72616/0b3f66bdd7b0c2090626a15566050f9f6d8351f2/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/d8a235b78557c3f731352b83b490f13e50fa90de/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: tests/queries/0_stateless/03277_empty_tuple_formats.sh [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+do
+  echo $format
+  $CLICKHOUSE_LOCAL -q "
+    insert into function file('$FILE', '$format', 'x Tuple()') select () as x settings engine_file_truncate_on_insert=1;
+    select * from file('$FILE', '$format', 'x Tuple()');"
+done
+
+# Formats that don't support empty tuples.
+$CLICKHOUSE_LOCAL -q "
+  insert into function file('$FILE', 'Parquet') select () settings engine_file_truncate_on_insert=1; -- {serverError BAD_ARGUMENTS}
```
Comment (antaljanosbenjamin) [inline tests/queries/0_stateless/03277_empty_tuple_formats.sh:40]: How does it come that this results in `BAD_ARGUMENTS`, while the similar query in `02735_parquet_encoder.sql` returns `PARQUET_EXCEPTION`?

Comment (antaljanosbenjamin) [reply tests/queries/0_stateless/03277_empty_tuple_formats.sh:40]: Oh, `02735_parquet_encoder.sql` failed:
```
024-11-29 03:46:31 Expected server error code: 723 but got: 36 (query: insert into function file(empty_tuple_02735.parquet) select (); -- {serverError PARQUET_EXCEPTION}).
2024-11-29 03:46:31 Received exception from server (version 24.12.1):
2024-11-29 03:46:31 Code: 36. DB::Exception: Received from localhost:9000. DB::Exception: Parquet doesn't support empty tuples. (BAD_ARGUMENTS)
2024-11-29 03:46:31 (query: insert into function file(empty_tuple_02735.parquet) select (); -- {serverError PARQUET_EXCEPTION})
```
So I guess that should be changed.

**Comment by antaljanosbenjamin** [comment]
And I think this is a noticeable bugfix, so it makes sense to have a changelog entry.

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**